### ISSUE TITLE
services.openssh.extraConfig: "Match" configuration may be prepended, sshd fails

### ISSUE THREAD

**Issue by makefu** [NixOS/nixpkgs#12265]
When configuration a Match block with openssh.extraConfig it may happen that this extra configuration is getting merged before the rest of the config is added:

```
  services.openssh.enable = true;
  services.openssh.extraConfig = ''
    Match User share
      ChrootDirectory /media
      ForceCommand internal-sftp
      AllowTcpForwarding no
      PermitTunnel no
      X11Forwarding no
  '';
```

In some cases this results in the Match block to be prepended which is a real problem for sshd because there is no way to close a match block.

To solve this one could either use a new variable without implicit merging by nix like this:

```
pkgs.writeText "sshd-config" ''
PIDFile 
all-the-other-stuff
...
${cfg.extraConfig}
```

or it may be possible to introduce an extra option for Building `Match` entries:

```
services.openssh.match = { "User share" = ''my configuration'' };
```

Additionally i think it would be nice to test the configuration after being written with `sshd -tf <config>` and only switch to the configuration if the test succeeded. `pkgs.runCommand` could be used for this, a friend showed me https://github.com/zalora/microgram/blob/dc5adff60602b2e97d20ae942f4c4198655c0395/pkgs/default.nix#L51-L63 which does something similar.

I really hate being locked out from machines.Giving the sshd a safety net before crashing may be a good idea.

**Comment by rnhmjoj** [comment]
I had the exact same problem. After some research I found out you can "close" the block with `Match all`.

**Comment by makefu** [comment]
Seems to be fixed by d9d6a92 . #13393 was a dupe of this issue.
